File: 2022/08.ts
const TREES = (await Deno.readTextFile("./inputs/08.txt"))
  .split("\n")
  .map((row) => row.split("").map((t) => Number(t)));

const SIZE = TREES.length;
let VISIBLE = 4 * SIZE - 4;
let SCENIC = 0;

const getHorizontalVision = (
  tree: number,
  x: number,
  start: number,
  end: number,
) => {
  for (let i = start; i < end; i++) {
    if (TREES[x][i] >= tree) return false;
  }
  return true;
};

const getVerticalVision = (
  tree: number,
  y: number,
  start: number,
  end: number,
) => {
  for (let i = start; i < end; i++) {
    if (TREES[i][y] >= tree) return false;
  }
  return true;
};

const getScenicScore = (x: number, y: number) => {
  const score = [0, 0, 0, 0];
  for (let i = x - 1; i >= 0; i--) {
    if (TREES[i][y] >= TREES[x][y]) {
      score[0]++;
      break;
    } else {
      score[0]++;
    }
  }
  for (let i = y - 1; i >= 0; i--) {
    if (TREES[x][i] >= TREES[x][y]) {
      score[1]++;
      break;
    } else {
      score[1]++;
    }
  }
  for (let i = y + 1; i < SIZE; i++) {
    if (TREES[x][i] >= TREES[x][y]) {
      score[2]++;
      break;
    } else {
      score[2]++;
    }
  }
  for (let i = x + 1; i < SIZE; i++) {
    if (TREES[i][y] >= TREES[x][y]) {
      score[3]++;
      break;
    } else {
      score[3]++;
    }
  }
  return score.filter(Boolean).reduce((a, b) => a * b, 1);
};

for (let x = 1; x < SIZE - 1; x++) {
  for (let y = 1; y < SIZE - 1; y++) {
    if (
      getHorizontalVision(TREES[x][y], x, 0, y) ||
      getHorizontalVision(TREES[x][y], x, y + 1, SIZE) ||
      getVerticalVision(TREES[x][y], y, 0, x) ||
      getVerticalVision(TREES[x][y], y, x + 1, SIZE)
    ) {
      VISIBLE++;
    }
  }
}

for (let x = 0; x < SIZE; x++) {
  for (let y = 0; y < SIZE; y++) {
    const score = getScenicScore(x, y);
    if (score > SCENIC) {
      SCENIC = score;
    }
  }
}

console.log(VISIBLE);
console.log(SCENIC);

File: 2022/08.py
import numpy as np


def get_score(trees, tree_height):
    if all(tree_height > tree for tree in trees):
        return len(trees)
    for index, value in enumerate(trees):
        if value >= tree_height:
            return index + 1


def part1(file):
    lines = file.split("\n")
    forest_height = len(lines)
    forest_width = len(lines[0])
    forest = np.array(list(file.replace("\n", "")), "uint8").reshape(
        forest_height, forest_width
    )
    tree_height = -1
    trees_visible = 2 * forest_height + 2 * forest_width - 4
    for row in range(1, forest_height - 1):
        for col in range(1, forest_width - 1):
            tree_height = forest[row][col]

            up = forest[:, col][0:row]
            if tree_height > np.amax(up):
                trees_visible += 1
                continue

            down = forest[:, col][row:][1:]
            if tree_height > np.amax(down):
                trees_visible += 1
                continue

            left = forest[row][0:col]
            if tree_height > np.amax(left):
                trees_visible += 1
                continue

            right = forest[row][col:][1:]
            if tree_height > np.amax(right):
                trees_visible += 1
                continue
    return trees_visible


def part2(file):
    lines = file.split("\n")
    forest_height = len(lines)
    forest_width = len(lines[0])
    forest = np.array(list(file.replace("\n", "")), "uint8").reshape(
        forest_height, forest_width
    )

    max_score = -1
    for row in range(0, forest_height):
        for col in range(0, forest_width):
            tree_height = forest[row][col]
            score = 1

            up = np.flip(forest[:, col][0:row])
            score *= get_score(up, tree_height)

            down = forest[:, col][row:][1:]
            score *= get_score(down, tree_height)

            left = np.flip(forest[row][0:col])
            score *= get_score(left, tree_height)

            right = forest[row][col:][1:]
            score *= get_score(right, tree_height)

            max_score = max(score, max_score)
    return max_score


if __name__ == "__main__":
    file = open("inputs/08.txt").read()
    print(part1(file))
    print(part2(file))
